blog/2023-11-01-feature-flags-sdks-architectures.md

```diff
@@ -0,0 +1,127 @@
+* First requests are not in cache, thus slower and causing network traffic.
+* Not every request can be efficiently cached, as some evaluations can be dynamic according to context with different values.
+* When there is no background task for getting real-time configuration changes event and refreshing the cache, values can take time to be updated.
+
+### Local evaluation
+
+With this approach, the feature flag configuration is saved locally at the SDK.  
+Feature flag evaluation is done locally via the SDK and not affecting network traffic.
+With this approach, configuration can be fetched on initialization, and be refreshed periodically and/or when there is a configuration change, which can be triggered at the SDK via stream event from the feature flag service. 
+Some SDKs sync data via WebSockets or [SSE](https://en.wikipedia.org/wiki/Server-sent_events).
+When this approach is used, evaluation analytics metrics can be aggregated at the SDK side and published separately, as not every flag evaluation triggers a message to the Features Service / Proxy.
+
+### Local evaluation - example flow
+
+<img src={require('@site/static/img/blog/feature-flags-sdks-architectures/Local-evaluation.png').default} />
+
+#### Advantages
+
+* Functional advantage is that intermediate issue like network or feature flag service errors are automatically handled, as configuration is cached, and flags are evaluated locally.
+  If an issue like network error is ongoing for a long time period, SDK provider state can also be reflected as stale.
+* Another functional advantage, initialization can have a fallback mechanism to dealing with situations like temporary network failure by starting from last received configuration or some other configuration, thus microservice can start with last feature flag configuration.
+* Feature flag configuration can be cached on CDN.
+* Feature flag evaluation is done locally via the SDK and not affecting network traffic, hence can be fast and efficient.
+* Working without a relay proxy can still be efficient from SDK point of view, as flags are evaluated locally and network traffic does not block common flow.
+
+#### Disadvantages
+
+* When feature flag configuration is very large with many feature flags, network traffic and local cache sizes can be increased.
+* Local evaluation should be implemented at the SDK, developed and maintained for multiple languages. Considering rules like context properties based and percentage rollout strategies via hashing to buckets, it can be hard to keep 100% feature parity across languages.
+
+#### A note on client-side SDKs for web browser / client app
+
+This approach is less suitable for client-side SDKs from a security point of view, as full feature flag configuration and authorization keys are exposed to end users.
```

Comment:
There's some debate on this and some vendors do send flag configurations to the client. I would recommend removing this section for now. Perhaps we could dive into this in more detail in a follow-up blog.

Reply:
I saw a [reference](https://developer.harness.io/docs/feature-flags/ff-sdks/sdk-overview/client-side-and-server-side-sdks/#:~:text=Server%2Dside%20SDKs%20download%20all,and%20store%20them%20in%20memory.&text=Client%2Dside%20SDK%20has%20a,and%20avoids%20most%20Harness%20communication.) which explains it:
> Security
The SDKs are embedded in your applications and run on the browser or on mobile devices. Therefore, these SDKs can be compromised by users when unpacking a mobile app or using the browser's developer tools to inspect the page. You should not use a server-side SDK key in client-side applications. Client-side SDKs only perform an evaluation of a flag and receive the result. They don't store all data about a flag.

I think it is worth mentioning so readers can at least be aware of some notable difference to client SDKs.
Maybe you want to rephrase it to reduce the debate issue ?

Reply:
I would recommend avoiding this in general since the blog is specifically focused on server-side SDKs. I do think this is an interesting topic but it requires a deeper dive to truly weigh the pros/cons of each approach. For example, if your rulesets don't contain sensitive information, it may be preferred to send the ruleset instead of sending PII to an external service.

Reply:
ok, removing it.